File: src/lib/archive.ts
import { getDb } from "@/lib/db";
import { plans, appSettings } from "@/lib/db/schema";
import { eq, and, isNull, lte } from "drizzle-orm";

function getSetting(key: string, defaultValue: string): string {
  const db = getDb();
  const setting = db
    .select()
    .from(appSettings)
    .where(eq(appSettings.key, key))
    .get();
  return setting?.value || defaultValue;
}

export function archiveCompletedPlans(): number {
  const db = getDb();
  const archiveDays = parseInt(getSetting("archive_after_days", "30"), 10);
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - archiveDays);
  const cutoffStr = cutoff.toISOString();

  const result = db
    .update(plans)
    .set({ archivedAt: new Date().toISOString() })
    .where(
      and(
        eq(plans.status, "completed"),
        isNull(plans.archivedAt),
        lte(plans.updatedAt, cutoffStr)
      )
    )
    .run();

  return result.changes;
}

export function cleanupArchivedPlans(): number {
  const db = getDb();
  const cleanupDays = parseInt(getSetting("cleanup_after_days", "90"), 10);
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - cleanupDays);
  const cutoffStr = cutoff.toISOString();

  const toDelete = db
    .select({ id: plans.id })
    .from(plans)
    .where(
      and(
        lte(plans.archivedAt, cutoffStr)
      )
    )
    .all();

  for (const plan of toDelete) {
    db.delete(plans).where(eq(plans.id, plan.id)).run();
  }

  return toDelete.length;
}

File: src-tauri/src/routes/archive.rs
use axum::{
    extract::State,
    Json,
};
use serde_json::{json, Value};

use crate::state::AppState;

pub async fn archive(State(state): State<AppState>) -> Json<Value> {
    let db = state.db.lock().unwrap();

    let archived = archive_completed_plans(&db);
    let cleaned = cleanup_archived_plans(&db);

    Json(json!({"archived": archived, "cleaned": cleaned}))
}

fn get_setting(db: &rusqlite::Connection, key: &str, default: &str) -> String {
    db.query_row(
        "SELECT value FROM app_settings WHERE key = ?1",
        rusqlite::params![key],
        |row| row.get::<_, String>(0),
    )
    .unwrap_or_else(|_| default.to_string())
}

fn archive_completed_plans(db: &rusqlite::Connection) -> i64 {
    let archive_days: i64 = get_setting(db, "archive_after_days", "30")
        .parse()
        .unwrap_or(30);

    let cutoff_sql = format!("datetime('now', '-{} days')", archive_days);
    let sql = format!(
        "UPDATE plans SET archived_at = datetime('now') WHERE status = 'completed' AND archived_at IS NULL AND updated_at <= {}",
        cutoff_sql
    );

    db.execute(&sql, []).unwrap_or(0) as i64
}

fn cleanup_archived_plans(db: &rusqlite::Connection) -> i64 {
    let cleanup_days: i64 = get_setting(db, "cleanup_after_days", "90")
        .parse()
        .unwrap_or(90);

    let cutoff_sql = format!("datetime('now', '-{} days')", cleanup_days);

    // Get plans to delete
    let sql = format!(
        "SELECT id FROM plans WHERE archived_at IS NOT NULL AND archived_at <= {}",
        cutoff_sql
    );
    let mut stmt = db.prepare(&sql).unwrap();
    let ids: Vec<String> = stmt
        .query_map([], |row| row.get(0))
        .unwrap()
        .filter_map(|r| r.ok())
        .collect();

    for id in &ids {
        db.execute("DELETE FROM plans WHERE id = ?1", rusqlite::params![id])
            .ok();
    }

    ids.len() as i64
}
